src/test/java/com/android/volley/toolbox/BasicNetworkTest.java

```diff
@@ -192,6 +193,43 @@ public void noConnection() throws Exception {
         verify(mMockRetryPolicy, never()).retry(any(VolleyError.class));
     }
 
+    @Test
+    public void noConnectionRetry() throws Exception {
+        MockHttpStack mockHttpStack = new MockHttpStack();
+        mockHttpStack.setExceptionToThrow(new IOException());
+        BasicNetwork httpNetwork = new BasicNetwork(mockHttpStack);
+        Request<String> request = buildRequest();
+        request.setRetryPolicy(mMockRetryPolicy);
+        request.setShouldRetryConnectionErrors(true);
+        doThrow(new VolleyError()).when(mMockRetryPolicy).retry(any(VolleyError.class));
```

Comment:
Why do we need to throw a VolleyError?

Reply:
That was in the original test to make sure that it did not retry on a NoConnectionError by default. It looks like it's in basically every test for the network. What is the main benefit of using assertThrows instead?

Reply:
assertThrows replaces the entire try / catch, so that's a one line assertion that makes the test clearer, essentially.

This was a different question though, I am not fully sure to understand why we throw an error on retry. Is it not to retry? @jpd236 probably has more context on it.

Reply:
Oh okay, that makes sense. I believe our current tests don't use JUnit5, and assertThrows isn't supported before that.

And yeah, I'm not sure about that.

Reply:
The RetryPolicy interface is to have retry throw the provided error if the retry could not be performed:

https://github.com/google/volley/blob/d1a3d5388c79ff1a6fdb904daeff9b39d0bb7d26/src/main/java/com/android/volley/RetryPolicy.java#L55

I'm not certain, but at a glance, I suspect that throwing is simply the easiest way to write this test, because if you don't throw, then Volley will attempt the request a second time. So you'd need to configure the mock chain to handle that case (and, in particular, to not fail the second time or else it would recurse indefinitely). Throwing still ends the request while allowing you to verify that the retry handler was indeed called.

Reply:
RetryPolicy throws an error to indicate that no more retries are available, it just returns void to retry. For this test, if it didn't throw at some point, the test would loop with infinite retries.